### PR TITLE
Added JSON type to MySQL and PostgreSQL schema

### PR DESCRIPTION
**Pull request by zhukovra**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #13019, #15247  

This is fixed and tests enriched PR #13783.
Correct realization of PR #15266.

**Status**: working on tests stabilization under new versions of DBs.

### CODE REVIEW

**Comment by samdark** [comment]
@SilverFire I think you'd like to take a look at it.

**Comment by zhukovra** [comment]
@samdark FYI I want to separate that PR with excluding "Fixed default values loaded from schema for PostgreSQL" commit because I fix it for test passing on PostgreSQL 9.6.

Another **important** question for me is why Travis tests works on outdated and rarely used versions of databases servers? Where can I read about that decision policy?

**Comment by samdark** [comment]
Travis works with whatever is provided by default. We can merge adjusted travis.yml if needed.

**Comment by samdark** [comment]
@zhukovra would you please merge with master? Also... what's left for the pull request to be finished?

**Comment by zhukovra** [comment]
@samdark I'll try finish PR today or tomorrow.

**Comment by SilverFire** [comment]
Please pay attention to changes introduced by https://github.com/yiisoft/yii2/pull/15348

**Comment by zhukovra** [comment]
I've found [a bug](https://bugs.php.net/bug.php?id=70384) with JSON column type in MySQL that will not be fixed in PHP < 5.6. Failed test is [here](https://travis-ci.org/zhukovra/yii2/jobs/337449017).

What do I need to do with it? Should I use `version_compare()` for making that test pass?

**Comment by SilverFire** [comment]
Yes, throw exception and leave a `TODO` comment to drop this check in Yii 2.1

**Comment by zhukovra** [comment]
@SilverFire @samdark please review.

File: .travis.yml [hunk 1]
```diff
@@ -49,6 +49,15 @@ cache:
 # try running against postgres 9.6
 addons:
   postgresql: "9.6"
+  apt:
+    sources:
+      - mysql-5.7-trusty
+    packages:
+      - libmysqlclient-dev
+      - libmysqlclient20
```
Comment (samdark) [inline .travis.yml]: Are both libraries necessary?

File: .travis.yml [hunk 2]
```diff
@@ -49,6 +49,15 @@ cache:
...
+      - mysql-community-client
```
Comment (samdark) [inline .travis.yml]: Why do we need a client?

File: .travis.yml [hunk 3]
```diff
@@ -76,10 +85,14 @@ matrix:
...
           repo_token: 2935307212620b0e2228ab67eadd92c9f5501ddb60549d0d86007a354d56915b
         apt:
+          sources:
+            - mysql-5.7-trusty
           packages:
-            - mysql-server-5.6
-            - mysql-client-core-5.6
-            - mysql-client-5.6
+            - libmysqlclient-dev
+            - libmysqlclient20
```
Comment (samdark) [inline .travis.yml]: See above.

File: .travis.yml [hunk 4]
```diff
@@ -91,10 +104,14 @@ matrix:
...
         postgresql: "9.6"
         apt:
+          sources:
+            - mysql-5.7-trusty
           packages:
-            - mysql-server-5.6
-            - mysql-client-core-5.6
-            - mysql-client-5.6
+            - libmysqlclient-dev
+            - libmysqlclient20
```
Comment (samdark) [inline .travis.yml]: See above.

File: framework/db/SchemaBuilderTrait.php [hunk 5]
```diff
@@ -272,4 +272,25 @@ public function money($precision = null, $scale = null)
 
         return $this->getDb()->getSchema()->createColumnSchemaBuilder(Schema::TYPE_MONEY, $length);
     }
+
+    /**
+     * Creates a json column.
```
Comment (samdark) [inline framework/db/SchemaBuilderTrait.php]: json -> JSON

File: framework/db/SchemaBuilderTrait.php [hunk 6]
```diff
@@ -272,4 +272,25 @@ public function money($precision = null, $scale = null)
...
+     * @return ColumnSchemaBuilder the column instance which can be further customized.
+     * @since 2.0.14
+     * @throws \yii\base\Exception
+     */
+    public function json()
+    {
+        /**
```
Comment (samdark) [inline framework/db/SchemaBuilderTrait.php]: Since it's not a phpdoc block, use `/*` with single `*`.

File: framework/db/SchemaBuilderTrait.php [hunk 7]
```diff
@@ -272,4 +272,25 @@ public function money($precision = null, $scale = null)
...
+         * TODO Remove in Yii 2.1
+         *
+         * Disabled due bug in MySQL extension
```
Comment (samdark) [inline framework/db/SchemaBuilderTrait.php]: due to bug

File: tests/framework/db/SchemaTest.php [hunk 8]
```diff
@@ -249,7 +249,7 @@ public function testGetPDOType()
 
     public function getExpectedColumns()
     {
-        return [
+        $columns = [
```
Comment (samdark) [inline tests/framework/db/SchemaTest.php]: returning it right away was fine.

File: tests/framework/db/mysql/QueryBuilderTest.php [hunk 9]
```diff
@@ -62,7 +62,23 @@ public function columnTypes()
                 $this->primaryKey()->comment('test')->after('col_before'),
                 "int(11) NOT NULL AUTO_INCREMENT PRIMARY KEY COMMENT 'test' AFTER `col_before`",
             ],
-        ]);
+        ];
+
+        /**
```
Comment (samdark) [inline tests/framework/db/mysql/QueryBuilderTest.php]: Use `/*`

**Review by samdark — APPROVED** [review]
Generally code is fine. There are minor issues to fix and questions to answer. Also a line for changelog is needed.

**Comment by samdark** [comment]
As I understand, this PR fixes everything that #15266 was going to fix, right?

File: .travis.yml (hunk 1 above)
Comment (zhukovra) [reply .travis.yml]: Nope. Copypaste from outdated Travis instruction. Fixed.

**Comment by zhukovra** [comment]
@samdark I think that bugs #15266 was fixed by @SilverFire. PR updated. Thank you for review.

**Review by SilverFire — APPROVED** [review]
Looks good to me

**Comment by samdark** [comment]
A line for changelog is needed.

**Comment by samdark** [comment]
Merged. Thank you!

**Comment by titanproger** [comment]
Hello every one! 
I have a question about new json support on PostgreSQL.
Is it way to create a json field by migration now?
I mean not a "jsonb" but "json"  type.
example 
```
$this->createTable('table_name', [
            "id" => $this->bigPrimaryKey(),
            "body" => "json",   // warning "json" generates jsonb type
        ]);
```
First my think - is create own QueryBuilder width $typeMap = [
   ....
   "json_raw"  => 'json'
]
and use json_raw abstraction type.

**Comment by finnan444** [comment]
@titanproger Hi, i also encountered this issue, may be you created a new issue?

**Pull request merged**